File: src/core/git.ts
import { execSync } from 'child_process';
import { GitError } from '../utils/errors.js';

/**
 * Build git clone command with proper quoting and flags
 * Removes shell command separators as defense-in-depth
 * IMPORTANT: Input validation SHOULD occur before calling this function
 * @param url - Repository URL to clone
 * @param targetPath - Destination path for clone
 * @returns Git clone command string
 */
export function buildGitCloneCommand(url: string, targetPath: string): string {
  // Remove semicolons as basic injection prevention (defense-in-depth)
  const safeUrl = url.replace(/;/g, '');
  const safePath = targetPath.replace(/;/g, '');
  return `git clone --quiet "${safeUrl}" "${safePath}"`;
}

/**
 * Build git checkout command with proper quoting and flags
 * Removes shell command separators as defense-in-depth
 * @param commit - Commit hash to checkout
 * @returns Git checkout command string
 */
export function buildGitCheckoutCommand(commit: string): string {
  // Remove semicolons as basic injection prevention
  const safeCommit = commit.replace(/;/g, '');
  return `git checkout --quiet "${safeCommit}"`;
}

/**
 * Build git fetch command with proper flags
 * @returns Git fetch command string
 */
export function buildGitFetchCommand(): string {
  return 'git fetch --quiet --all';
}

/**
 * Clone a git repository
 * @param url - Repository URL to clone
 * @param targetPath - Destination path for clone
 * @throws GitError if clone fails
 */
export function gitClone(url: string, targetPath: string): void {
  try {
    const command = buildGitCloneCommand(url, targetPath);
    execSync(command, {
      encoding: 'utf-8',
      stdio: 'pipe',
      windowsHide: true
    });
  } catch (error) {
    const err = error as Error & { stderr?: string };
    const errorMessage = err.stderr || err.message;
    throw new GitError(`Failed to clone ${url}: ${errorMessage}`);
  }
}

/**
 * Checkout a specific commit in a git repository
 * @param repoPath - Path to the git repository
 * @param commit - Commit hash to checkout
 * @throws GitError if checkout fails
 */
export function gitCheckout(repoPath: string, commit: string): void {
  try {
    const command = buildGitCheckoutCommand(commit);
    execSync(command, {
      cwd: repoPath,
      encoding: 'utf-8',
      stdio: 'pipe',
      windowsHide: true
    });
  } catch (error) {
    const err = error as Error & { stderr?: string };
    const errorMessage = err.stderr || err.message;
    throw new GitError(`Failed to checkout commit ${commit}: ${errorMessage}`);
  }
}

/**
 * Fetch updates from remote repository
 * @param repoPath - Path to the git repository
 * @throws GitError if fetch fails
 */
export function gitFetch(repoPath: string): void {
  try {
    const command = buildGitFetchCommand();
    execSync(command, {
      cwd: repoPath,
      encoding: 'utf-8',
      stdio: 'pipe',
      windowsHide: true
    });
  } catch (error) {
    const err = error as Error & { stderr?: string };
    const errorMessage = err.stderr || err.message;
    throw new GitError(`Failed to fetch updates: ${errorMessage}`);
  }
}

/**
 * Get the current commit hash of a git repository
 * @param repoPath - Path to the git repository
 * @returns Current commit hash (40 characters)
 * @throws GitError if unable to get current commit
 */
export function getCurrentCommit(repoPath: string): string {
  try {
    const output = execSync('git rev-parse HEAD', {
      cwd: repoPath,
      encoding: 'utf-8',
      stdio: 'pipe',
      windowsHide: true
    });
    return output.trim();
  } catch (error) {
    const err = error as Error & { stderr?: string };
    const errorMessage = err.stderr || err.message;
    throw new GitError(`Failed to get current commit: ${errorMessage}`);
  }
}

File: lua/necromancer/core/git.lua
local errors = require("necromancer.utils.errors")

local M = {}

---Execute a git command synchronously
---@param args string[] Git command arguments
---@param cwd? string Working directory
---@return string output
local function git_exec(args, cwd)
  local cmd = vim.list_extend({ "git" }, args)

  local result
  if cwd then
    -- Change to directory, run command, change back
    local original_dir = vim.fn.getcwd()
    vim.fn.chdir(cwd)
    result = vim.fn.system(cmd)
    vim.fn.chdir(original_dir)
  else
    result = vim.fn.system(cmd)
  end

  if vim.v.shell_error ~= 0 then
    error(errors.GitError("Git command failed: " .. vim.trim(result)))
  end

  return vim.trim(result)
end

---Clone a git repository
---@param url string Repository URL
---@param target_path string Destination path
function M.clone(url, target_path)
  -- Basic injection prevention (defense in depth)
  local safe_url = url:gsub(";", "")
  local safe_path = target_path:gsub(";", "")

  local ok, err = pcall(function()
    git_exec({ "clone", "--quiet", safe_url, safe_path })
  end)

  if not ok then
    local msg = type(err) == "table" and err.message or tostring(err)
    error(errors.GitError("Failed to clone " .. url .. ": " .. msg))
  end
end

---Checkout a specific commit
---@param repo_path string Path to repository
---@param commit string Commit hash
function M.checkout(repo_path, commit)
  local safe_commit = commit:gsub(";", "")

  local ok, err = pcall(function()
    git_exec({ "checkout", "--quiet", safe_commit }, repo_path)
  end)

  if not ok then
    local msg = type(err) == "table" and err.message or tostring(err)
    error(errors.GitError("Failed to checkout " .. commit .. ": " .. msg))
  end
end

---Fetch updates from remote
---@param repo_path string Path to repository
function M.fetch(repo_path)
  local ok, err = pcall(function()
    git_exec({ "fetch", "--quiet", "--all" }, repo_path)
  end)

  if not ok then
    local msg = type(err) == "table" and err.message or tostring(err)
    error(errors.GitError("Failed to fetch: " .. msg))
  end
end

---Get current commit hash
---@param repo_path string Path to repository
---@return string commit 40-character hash
function M.get_current_commit(repo_path)
  local ok, result = pcall(function()
    return git_exec({ "rev-parse", "HEAD" }, repo_path)
  end)

  if not ok then
    local msg = type(result) == "table" and result.message or tostring(result)
    error(errors.GitError("Failed to get current commit: " .. msg))
  end

  return result
end

---Get default branch name from remote
---@param repo_path string Path to repository
---@return string branch Default branch name
function M.get_default_branch(repo_path)
  -- Try to get from remote
  local ok, result = pcall(function()
    return git_exec({ "remote", "show", "origin" }, repo_path)
  end)

  if ok and result then
    local branch = result:match("HEAD branch: ([^\n]+)")
    if branch and #branch > 0 then
      return vim.trim(branch)
    end
  end

  -- Fallback: check if main or master exists
  local main_ok = pcall(function()
    git_exec({ "rev-parse", "--verify", "main" }, repo_path)
  end)
  if main_ok then
    return "main"
  end

  local master_ok = pcall(function()
    git_exec({ "rev-parse", "--verify", "master" }, repo_path)
  end)
  if master_ok then
    return "master"
  end

  -- Default to main
  return "main"
end

return M
